File: golang-for-nodejs-developers/examples/streams.js
import { Readable, Writable } from 'node:stream';

const inStream = new Readable({
  read() {}
});

inStream.push(Buffer.from('foo'));
inStream.push(Buffer.from('bar'));
inStream.push(null); // end stream
inStream.pipe(process.stdout);

// Add a newline for clarity in output
inStream.on('end', () => console.log());

const outStream = new Writable({
  write(chunk, encoding, callback) {
    console.log(`received: ${chunk.toString('utf8')}`);
    callback();
  }
});

outStream.write(Buffer.from('abc'));
outStream.write(Buffer.from('xyz'));
outStream.end();

File: golang-for-nodejs-developers/examples/streams.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"os"
	"time"
)

func main() {
	inStream := new(bytes.Buffer)
	w := bufio.NewWriter(inStream)
	_, err := w.Write([]byte("foo"))
	if err != nil {
		panic(err)
	}
	_, err = w.Write([]byte("bar"))
	if err != nil {
		panic(err)
	}
	err = w.Flush()
	if err != nil {
		panic(err)
	}

	inStream.WriteTo(os.Stdout)
	fmt.Println()

	outStream := new(bytes.Buffer)
	outStream.Write([]byte("abc\n"))
	outStream.Write([]byte("xyz\n"))
	piper, pipew := io.Pipe()

	go func() {
		sc := bufio.NewScanner(piper)
		for sc.Scan() {
			fmt.Println("received: " + sc.Text())
		}
		if err := sc.Err(); err != nil {
			panic(err)
		}
	}()

	go func() {
		defer pipew.Close()
		io.Copy(pipew, outStream)
	}()

	// Give it a bit of time to complete
	time.Sleep(100 * time.Millisecond)
}
